### PR TITLE
Update Alertmanager with support for UTF-8

### PR DESCRIPTION
**Pull request by grobinson-grafana**
This pull request updates our fork of Alertmanager to commit [65bdab0](https://github.com/grafana/prometheus-alertmanager/commit/65bdab09d3ed64c7bdd9474f19d4d48f857240fb), which is based on commit [5658f8c](https://github.com/prometheus/alertmanager/commit/5658f8cfb59c9e23bcfed3bc66cd47a7b6d92a85) in Prometheus Alertmanager. It removes the overrides for validation of alerts, labels and silences as UTF-8 is now supported upstream. The compat package, which controls the enablement of UTF-8, is initialized in fallback mode using the noop flags.

### CODE REVIEW

File: notify/alerts.go [hunk 1]
```diff
@@ -220,7 +221,7 @@ func parseReceivers(receivers string) (*regexp.Regexp, error) {
 func parseFilter(filter []string) ([]*labels.Matcher, error) {
 	matchers := make([]*labels.Matcher, 0, len(filter))
 	for _, matcherString := range filter {
-		matcher, err := labels.ParseMatcher(matcherString)
+		matcher, err := compat.Matcher(matcherString, "unknown")
```
Comment (grobinson-grafana) [inline notify/alerts.go]: This is called from `GetAlerts` and `GetAlertGroups` in `grafana/grafana` which in most cases is for API calls. I didn't see calls to these functions outside of the Alertmanager API or tests.

File: notify/alerts.go [hunk 2]
```diff
@@ -220,7 +221,7 @@ func parseReceivers(receivers string) (*regexp.Regexp, error) {
 func parseFilter(filter []string) ([]*labels.Matcher, error) {
 	matchers := make([]*labels.Matcher, 0, len(filter))
 	for _, matcherString := range filter {
-		matcher, err := labels.ParseMatcher(matcherString)
+		matcher, err := compat.Matcher(matcherString, "api")
```
Comment (grobinson-grafana) [inline notify/alerts.go:224]: This is called from `GetAlerts` and `GetAlertGroups` in `grafana/grafana` which in most cases is for API calls. I didn't see calls to these functions outside of the Alertmanager API or tests.

File: .golangci.yml [hunk 3]
```diff
@@ -4,10 +4,10 @@ output:
 linters:
   enable:
     - goimports
-    - golint
```
Comment (grobinson-grafana) [inline .golangci.yml:7]: `golint` is deprecated and superseded with `revive`. Lots of linter warnings with revive, which have been ignored. Issue to fix this is tracked [here](https://github.com/grafana/alerting-squad/issues/724).

File: notify/grafana_alertmanager.go [hunk 4]
```diff
@@ -511,7 +497,7 @@ func (am *GrafanaAlertmanager) PutAlerts(postableAlerts amv2.PostableAlerts) err
 			am.Metrics.Resolved().Inc()
 		}
 
-		if err := validateAlert(alert); err != nil {
+		if err := alert.Validate(); err != nil {
```
Comment (grobinson-grafana) [inline notify/grafana_alertmanager.go:505]: This is covered with existing tests [here](https://github.com/grafana/alerting/blob/1d58deb6074c6985fdc21e0e514a87a6f9e1dd3e/notify/grafana_alertmanager_test.go#L39).

File: notify/grafana_alertmanager.go [hunk 5]
```diff
@@ -45,21 +42,10 @@ const (
...
-		case pb.Matcher_REGEXP, pb.Matcher_NOT_REGEXP:
-			if _, err := regexp.Compile(m.Pattern); err != nil {
-				return fmt.Errorf("invalid regular expression %q: %s", m.Pattern, err)
-			}
-		default:
-			return fmt.Errorf("unknown matcher type %q", m.Type)
-		}
-		return nil
-	}
+	// This initializes the compat package in fallback mode. It parses first using the UTF-8 parser
```
Comment (grobinson-grafana) [inline notify/grafana_alertmanager.go:45]: The reason we need to use fallback mode to start with is that Grafana is still used to manage configurations in Mimir and Alertmanager. Until most installations are updated, we will need to operate in fallback mode. I don't think we can take a %, but instead change it after some period of time (like 6 months).

File: notify/grafana_alertmanager.go [hunk 6]
```diff
@@ -45,21 +42,10 @@ const (
...
+	// and then fallsback to the classic parser on error. UTF-8 is permitted in label names.
+	// This should be removed when the compat package is removed from Alertmanager.
+	compat.InitFromFlags(log.NewNopLogger(), compat.RegisteredMetrics, featurecontrol.NoopFlags{})
```
Comment (grobinson-grafana) [inline notify/grafana_alertmanager.go:48]: The reason we need to use fallback mode to start with is that Grafana can manage configurations in Cortex and Mimir, and read configurations from Prometheus Alertmanager. Until most of those installations are using UTF-8 mode, Grafana will need to operate in fallback mode. I don't think we can take a %, but instead change it after some period of time (like 6 months).

File: notify/alerts.go (hunk 2 above)
Comment (grobinson-grafana) [reply notify/alerts.go:224]: @grobinson-grafana to use the function in Alertmanager [here](https://github.com/prometheus/alertmanager/blob/f708f32ac7a3370d90a2bba326ff1f104816d518/api/v2/api.go#L677) and delete this.

**Review by gotjosh — APPROVED** [review]
LGTM

**Review by gotjosh — APPROVED** [review]
LGTM

**Pull request merged**